### ISSUE TITLE
Sendmail Installer plugin

### ISSUE THREAD

**Issue by TheNavigat** [certbot/certbot#6009]
Create a sendmail configurator.

The following is a *very* crude list of notes I have made while researching sendmail. The list will be updates as findings keep surfacing.

Everything here is nothing more than mere observations. Please doubt everything. Double check everything you rely on. I make a lot of mistakes so please don’t blindly rely on the information written here.

- sendmail.conf is a parameter file. It looks like it defines very simple key and value parameters, nothing beyond foo=”bar”. The default configuration file has all the keys in capital letters, snake-case (for example: THIS_IS_A_KEY) and all the values as quoted strings using double quotes. Didn’t check if this is a requirement.
- sendmail.cf is operating-system specific, automatically generated. It’s not made to be human readable.
- sendmail.mc is another real configuration file meant to be touched by humans. It generates sendmail.cf
- sendmail.cf remains the principal sendmail configuration file. Used for the Mail Transmission Agent (MTA).
- submit.cf is used for the Mail Submission Program (MSP). The MSP is used to submit mail messages. Unlike the MTA, it does not run as an SMTP daemon.
- I couldn’t find a good enough documentation so I resorted to downloading the source code from https://www.proofpoint.com/us/open-source-email-solution. The source code contains amazing README files. cf/README looks like the main file we'll need for documentation.
- Looks like the only file we actually need to edit/parse is `sendmail.mc`

### Sendmail.mc
- dnl is `delete through newline`.
- sourcecode/cf/README: Sendmail uses the M4 macro processor to "compile'' the configuration files.  The most important thing to know is that M4 is stream-based, that is, it doesn't understand about lines.  For this reason, in some places you may see the word "dnl'', which stands for "delete through newline''; essentially, it deletes all characters starting at the "dnl'' up to and including the next newline character.  In most cases sendmail uses this only to avoid lots of unnecessary blank lines in the output.

**Comment by jgillula** [comment]
Flagging our fearless starttls lead, @sydneyli

**Comment by sydneyli** [comment]
Thank you @TheNavigat!!!! This looks awesome-- thanks for working on this!

**Comment by TheNavigat** [comment]
Thanks @sydneyli! I didn't actually *start working on it* yet, that's just some research. I'll probably start with a `sendmail.mc` parser, unless someone is working on that already.

**Comment by sydneyli** [comment]
Sounds great. Let us know if you have any questions about it-- I'll take a look at Sendmail things/your notes some time this week also.

**Comment by pde** [comment]
I did some research on this when deciding whether Sendmail support was sane or feasible at all. What I was able to determine was that Debian has a bunch of scripts that modify/expand the sendmail config (which is Turing complete, of course) to provide a sane, simple STARTTLS config.

You can see this when you install Sendmail on a debian system:

```
Creating/Updating SSL(for TLS) information
Creating /etc/mail/tls/starttls.m4...
Creating SSL certificates for sendmail.
Generating DSA parameters, 2048 bit long prime
This could take some time
.................................+...+........+.+.+....+++++++++++++++++++++++++++++++++++++++++++++++++++*
..+..........+...+..+....+.........+...........+...................+.................................+.....+...............+......+..+....+......+............+.........+++++++++++++++++++++++++++++++++++++++++++++++++++*
Generating RSA private key, 2048 bit long modulus
..........................................................................+++
.........................................................+++
e is 65537 (0x010001)

*** *** *** WARNING *** WARNING *** WARNING *** WARNING *** *** ***
Everything you need to support STARTTLS (encrypted mail transmission
and user authentication via certificates) is installed and configured
but is *NOT* being used.

To enable sendmail to use STARTTLS, you need to:
1) Add this line to /etc/mail/sendmail.mc and optionally
   to /etc/mail/submit.mc:
  include(`/etc/mail/tls/starttls.m4')dnl
2) Run sendmailconfig
3) Restart sendmail
```

These outputs are (if I hazily recall correctly) a postinst script. It would probably be best to include a (possibly customized) version of that script in our plugin, and run it to allow the comparatively friendly `/etc/mail/tls/starttls.m4` file:

```
cat /etc/mail/tls/starttls.m4
divert(-1)dnl
####################################################################
##### This file is automagically generated -- edit at your own risk
#####
##### Copyright (c) 2002-2010 Richard Nelson.  All Rights Reserved.
#####
##### file: /etc/mail/tls/starttls.m4
#####		STARTTLS Configuration for Debian Sendmail
##### generated via: (/usr/bin/perl v5>26.1)
#####		/usr/share/sendmail/update_tlsm4
#####		version: 8.15.2 2018-03-20 00:49:16 cowboy
##### by: pde@graphene
##### on: Tue May 22 15:23:16 2018
##### in: /
##### input files: /etc/mail/databases
#####
##### Usage:
#####	1) To get *ANY* STARTTLS support for sendmail you
#####		A) *MUST* Add this line to /etc/mail/sendmail.mc
#####		   `include(`/etc/mail/tls/starttls.m4')dnl'
#####		B) *MAY* Add the same line to /etc/mail/submit.mc
#####		   to get MSP<->MTA authentication/encryption
#####	2) You may modify the marked portions of this file, those
#####	   deal with the cert/key filenames and TLS options.
#####	   If you need finer control of TLS options, use the access
#####	   database.
#####
####################################################################
divert(0)dnl
VERSIONID(`$Id: starttls.m4,v 8.15.2-11 2018-03-20 00:49:16 cowboy Exp $')
dnl #
dnl #---------------------------------------------------------------------
dnl # Bring in Autoconf results
dnl #---------------------------------------------------------------------
ifdef(`sm_version', `dnl',
`include(`/usr/share/sendmail/cf/debian/autoconf.m4')dnl')
dnl #
dnl # Check to see if inclusion is valid (version >= 8.11.0, tls enabled)
ifelse(eval(sm_version_math >= 527104), `1', `dnl
ifelse(sm_enable_tls, `yes', `dnl
dnl #
dnl # To support shared keyfiles, we need them to be group readable
dnl #
define(`confDONT_BLAME_SENDMAIL',dnl
	defn(`confDONT_BLAME_SENDMAIL')`,GroupReadableKeyFile')dnl
dnl #
dnl #             ...Do not touch anything above this line...
dnl #
dnl # Set a more reasonable timeout on negotiation
dnl #
define(`confTO_STARTTLS',     `2m')dnl   # <= EDIT
dnl #
dnl # CA directory - CA certs should be herein
define(`confCACERT_PATH',     `/etc/ssl/certs')dnl   # <= EDIT
dnl #
dnl # CA file (may be the same as client/server certificate)
define(`confCACERT',           `/etc/mail/tls/sendmail-server.crt')dnl   # <= EDIT
dnl #
dnl # Certificate Revocation List
define(`confCRL',              `')dnl   # <= EDIT
dnl # CRL not found...  do not issue warnings on it!
undefine(`confCRL')dnl
dnl #
dnl # Server certificate/key (can be in the same file, and shared w/client)
dnl # NOTE: The key must *NOT* be encrypted !!!
define(`confSERVER_CERT',     `/etc/mail/tls/sendmail-server.crt')dnl   # <= EDIT
define(`confSERVER_KEY',      `/etc/mail/tls/sendmail-common.key')dnl   # <= EDIT
dnl #
dnl # Client certificate/key (can be in the same file, and shared w/server)
dnl # NOTE: The key must *NOT* be encrypted !!!
define(`confCLIENT_CERT',     `/etc/mail/tls/sendmail-client.crt')dnl   # <= EDIT
define(`confCLIENT_KEY',      `/etc/mail/tls/sendmail-common.key')dnl   # <= EDIT
dnl #
dnl # DH parameters
define(`confDH_PARAMETERS',   `/etc/mail/tls/sendmail-common.prm')dnl # <= EDIT
dnl #
dnl # Optional settings
define(`confTLS_SRV_OPTIONS', `V')dnl   # <= EDIT
dnl #
')')dnl
```

**Comment by pde** [comment]
The Debian scripts in question appear to be these two, in the `sendmail-base` package:

```
/usr/share/sendmail/update_tls
/usr/share/sendmail/update_tlsm4
```
Would be good to take them and see how much modification is required to get them to work on a fairly different (eg FreeBSD :) ) system

**Comment by TheNavigat** [comment]
I've taken a look through these. I'm not a bash/perl expert so please correct me if any of the following statements is wrong.

It looks like these scripts first assert whether there's a link to `/etc/mail/tls/starttls.m4` and then through that file configure the certificates correctly.

While this (apparently) works, all our other configurations tend to take a slightly different approach. We analyze the currently existing configuration files and check for existing SSL configuration and either modify it, add it, or report an error.

On the other hand, it looks to me as if the `update_tls` scripts does not check for modifications in the current configuration files. This makes sense for them because these scripts are meant to run on installation and later used when the scripts have already been run before, but in our case I think we need to take our usual route.

tl;dr: We probably still need to correctly parse the existing configuration files for existing SSL configuration. Right?

**Comment by sydneyli** [comment]
@TheNavigat Thanks for looking into this. It looks like there's a huge complexity wall here. The robustness of parsing everything may be offset by the sheer complexity and overhead of maintaining the resulting codebase. How far can we get without writing a full parser+processor for m4 in Python? Can we do something simple to catch at least 90% of configuration setups? For the final 10% of use cases, we'd need to focus on detecting the potential failure, and either warning or prompting the user before continuing with further action-- can we do that reliably, too?

So, two main questions to mull over:
1. How many use cases do we catch by simply rolling our own version of `starttls.m4`, then dropping `include(starttls.m4)` into `sendmail.mc`?
2. How reliably can we guess that (1) will fail in some way?

And some side questions to think about later:
 -  What are the cases where a single Sendmail instance might want to install multiple certificates, and which of these cases should we handle?  Might someone want to host a submission server on a separate hostname from the MTA? Might someone want to install a cert corresponding with their mail domain(s) and not their hostname?
     - Background: any certificate installation via Certbot also knows about the `domain` for which that certificate is being issued. For HTTP servers, this is used to determine which "virtual host" to install the certificate on. For MTAs, this notion doesn't make as much sense. MTA hostnames and IPs are typically mapped 1-1-- most MTAs don't support SNI for that reason. However, sendmail does have a `DOMAIN` macro to specify per-domain parameters, but these seem to correspond with the mail domain (the part after the @ in an email address) and not the server's hostname.
 - And lastly, I'm curious whether other distros ship any additions to sendmail.

Hopefully the above was somewhat helpful? Let me know if you have any questions about the above. I won't have the bandwidth to work directly on this for a month or so, but am happy to continue investigating things like common configuration setups, usage edge cases, etc. Will let you know here if I learn anything relevant.

**Comment by TheNavigat** [comment]
The thing I'm missing is the reason why parsing m4 is difficult in the first place.

I can't remember the last time I worked with m4, but `sendmail.mc` and `starttls.mc` look fairly simple to parse. Where is the assumption that it will be that difficult coming from?

**Comment by pde** [comment]
@TheNavigat there's a whole [area of computer science](https://medium.com/@chetcorcos/introduction-to-parsers-644d1b5d7f3d) that explains why it's hard. But it boils down to the fact that those files can be modified to include arbitrary m4 code, and would remain valid.

**Comment by TheNavigat** [comment]
@pde Oh, I did actually study this, but for some reason my brain skipped it when I thought of parsing the configuration files. I didn't know that m4 is an actual complete language. Thanks a lot for explaining, now I get it...

**Comment by sydneyli** [comment]
yep, m4 itself is turing-complete and more fun: by accident, sendmail's config language that they compile to [is also](http://okmij.org/ftp/Computation/sendmail-as-turing-machine.txt) ¯\_(ツ)_/¯

**Comment by TheNavigat** [comment]
From what it looks like, the best approach would be to follow the footsteps of Debian, and then start figuring out solutions to issues as they rise. Should I give this a shot? What do you folks think?

**Comment by sydneyli** [comment]
Yes, I think that's reasonable. I'm also curious about one of the questions I posted above, namely, should we even do anything with the `domain` parameter given to the installer?

**Comment by sydneyli** [comment]
Okay. So @TheNavigat if you're still interested in working on this at all, there has been some confusion surrounding these plugins on our end. We've only recently been able to reach a point of clarity.

Originally, some folks had told us "We want Certbot to work with Dovecot & Sendmail." this message was then relayed to the team as "We want Certbot installer plugins for Dovecot & Sendmail" and motivated this work earlier this year. After several clarification conversations it turns out that they actually meant "We want Certbot to obtain certificates for our mailserver, but also we don't want Certbot to touch our mailserver configurations."

This is already possible. It's become abundantly clear that we don't have good documentation on best practices for using Certbot with non-webservers, and we should write more. I've thrown a quick guide for mailserver administration + Certbot up here, and may turn this into a pinned post on the community forums, or a blog post, or integrate into the Certbot docs, or some/all of those things.

As for the status of the Dovecot/Sendmail installers, we have continued to receive feedback from mailserver administrators that they do not want Certbot touching their config files. So we may not want full-fledged installers after all. I still think these would be useful, but we have deprioritized the actual installers for the time being.

That said, although Certbot + Dovecot behaves nicely, Certbot + Sendmail seems to have a couple of hiccups with regard to setup that we could make easier. I'm doing a separate writeup on this currently-- @TheNavigat, let me know if you would like to be looped in on this, and if you would be interested in helping out with this effort.

I'm sorry for the confusion and lack of communication on our part. We're also currently in discussion to set up a public communication channel with external developers so that we can be more transparent and responsive about these things in the future.

**Comment by sydneyli** [comment]
Thanks @TheNavigat !
> As you can see from my very erratic frequency of contribution due to university and other stuff (internships) the amount of time I can dedicate to Certbot varies highly, but I am still very interested and I would like to contribute further :)

That's understandable! I'll keep you looped in as this develops.

> Regarding your quick guide, can we possibly add this to https://certbot.eff.org ?

Yes, that's the idea! A lot of the documentation here might change shortly, as we're currently fixing up Certbot's packaging & installation situation. When that happens, we'll try to work in non-webserver guides as well.

**Comment by sydneyli** [comment]
> Certbot + Sendmail seems to have a couple of hiccups with regard to setup that we could make easier.

The most problematic hiccup is #1473, so I'll be working on that shortly and draft up an implementation proposal there.

Other than that installation seems to work fine so long as you modify the correct variables to point to fullchain and privkey. The least obvious one is that you also need to set `confCACERT` to point to a file that contains one of the certs in `fullchain`. So also pointing to `fullchain` works, but I'm not sure if it's best practice. From my understanding, this doesn't actually specify the certs that sendmail will use to validate incoming TLS connections; that's `confCACERT_PATH`, which should be set by default to your system's root CA store (often `/etc/ssl/certs`).

I've thrown up a janky PoC onto the `sendmail` branch if folks are interested in trying it. It has some workarounds for #1473 but otherwise doesn't do that much, given that we don't want to actually modify the config.

**Comment by sydneyli** [comment]
We've decided not to work on this. See #7097 !
